ovh/resource_cloud_project_kube_iprestriction.go

```diff
@@ -0,0 +1,129 @@
+	"net/url"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/ovh/terraform-provider-ovh/ovh/helpers"
+)
+
+func resourceCloudProjectKubeIpRestrictions() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceCloudProjectKubeIpRestrictionsCreate,
+		Read:   resourceCloudProjectKubeIpRestrictionsRead,
+		Delete: resourceCloudProjectKubeIpRestrictionsDelete,
+		Update: resourceCloudProjectKubeIpRestrictionsUpdate,
+
+		Importer: &schema.ResourceImporter{
+			State: resourceCloudProjectKubeIpRestrictionsImportState,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"service_name": {
+				Type:        schema.TypeString,
+				Description: "Service name",
+				Required:    true,
+				ForceNew:    true,
+				DefaultFunc: schema.EnvDefaultFunc("OVH_CLOUD_PROJECT_SERVICE", nil),
+			},
+			"kube_id": {
+				Type:        schema.TypeString,
+				Description: "Kube ID",
+				Required:    true,
+				ForceNew:    true,
+			},
+			"ips": {
+				Type:        schema.TypeSet,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Set:         schema.HashString,
+				Description: "List of ip restrictions for the cluster",
+				Required:    true,
+			},
+		},
+	}
+}
+
+func resourceCloudProjectKubeIpRestrictionsImportState(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	givenId := d.Id()
+	splitId := strings.SplitN(givenId, "/", 3)
+	if len(splitId) != 2 {
+		return nil, fmt.Errorf("Import Id is not service_name/kubeid formatted")
+	}
+	serviceName := splitId[0]
+	kubeId := splitId[1]
+	d.SetId(kubeId)
+	d.Set("kube_id", kubeId)
+	d.Set("service_name", serviceName)
+
+	results := make([]*schema.ResourceData, 1)
+	results[0] = d
+	return results, nil
+}
+
+func resourceCloudProjectKubeIpRestrictionsCreate(d *schema.ResourceData, meta interface{}) error {
+	return resourceCloudProjectKubeIpRestrictionsUpdate(d, meta)
+}
+
+func resourceCloudProjectKubeIpRestrictionsRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	serviceName := d.Get("service_name").(string)
+	kubeId := d.Get("kube_id").(string)
+
+	endpoint := fmt.Sprintf("/cloud/project/%s/kube/%s/ipRestrictions", serviceName, kubeId)
+	res := make(CloudProjectKubeIpRestrictionsResponse, 0)
+
+	log.Printf("[DEBUG] Will read iprestrictions from cluster %s in project %s", kubeId, serviceName)
+	if err := config.OVHClient.Get(endpoint, &res); err != nil {
+		return helpers.CheckDeleted(d, err, endpoint)
+	}
+
+	d.SetId(kubeId)
+	d.Set("ips", res)
+
+	log.Printf("[DEBUG] Read iprestrictions: %+v", res)
+	return nil
+}
+
+func resourceCloudProjectKubeIpRestrictionsUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	serviceName := d.Get("service_name").(string)
+	kubeId := d.Get("kube_id").(string)
+
+	endpoint := fmt.Sprintf("/cloud/project/%s/kube/%s/ipRestrictions", serviceName, kubeId)
+	params := (&CloudProjectKubeIpRestrictionsCreateOpts{}).FromResource(d)
```

Comment:
in types_cloud_project_kube_iprestrictiones, you have 2 types (create & update), 
here we use the CreateOpts type in the Update method., 
the create method points to the the udpdate methode

ihmo, it brings a lot of confustion.

either we have 2 explicit methods, with 2 explicit types
or we can have one method CreateOrUpdate with one type CreateOrUpdateOpts

this way, you can point to this only method in the schema definition.

what do you think ?

Reply:
I agree with you on this!
Fixed.